Keda.CosmosDbScaler/Services/CosmosDbMetricProvider.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.Azure.Cosmos;
+using Microsoft.Extensions.Logging;
+
+namespace Keda.CosmosDbScaler
+{
+    internal sealed class CosmosDbMetricProvider : ICosmosDbMetricProvider
+    {
+        private readonly CosmosDbFactory _factory;
+        private readonly ILogger<CosmosDbMetricProvider> _logger;
+
+        public CosmosDbMetricProvider(CosmosDbFactory factory, ILogger<CosmosDbMetricProvider> logger)
+        {
+            _factory = factory ?? throw new ArgumentNullException(nameof(factory));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task<long> GetPartitionCountAsync(ScalerMetadata scalerMetadata)
+        {
+            try
+            {
+                Container leaseContainer = _factory
+                    .GetCosmosClient(scalerMetadata.LeaseConnection)
+                    .GetContainer(scalerMetadata.LeaseDatabaseId, scalerMetadata.LeaseContainerId);
+
+                ChangeFeedEstimator estimator = _factory
+                    .GetCosmosClient(scalerMetadata.Connection)
+                    .GetContainer(scalerMetadata.DatabaseId, scalerMetadata.ContainerId)
+                    .GetChangeFeedEstimator(scalerMetadata.ProcessorName, leaseContainer);
+
+                // It does not help by creating more change-feed processing instances than the number of partitions.
+                int partitionCount = 0;
+
+                using (FeedIterator<ChangeFeedProcessorState> iterator = estimator.GetCurrentStateIterator())
+                {
+                    while (iterator.HasMoreResults)
+                    {
+                        FeedResponse<ChangeFeedProcessorState> states = await iterator.ReadNextAsync();
+                        partitionCount += states.Where(state => state.EstimatedLag > 0).Count();
```

Comment:
Why only counting the number of states that have lag > 0? Each state represents 1 lease, so the count of the partitions is = to the number of states regardless of current lag

Reply:
@ealsur, though the variable name is `partitionCount`, it is the count of partitions that have one or more items pending to be processed. The value returned by the method governs how many hosts will be spawned as in [this diagram](https://docs.microsoft.com/en-us/azure/cosmos-db/sql/change-feed-processor#components-of-the-change-feed-processor). For example, if number of partitions is 2, but there are no items pending for consumption, the method should return 0.

I created a test-application to confirm that number of states would be non-zero even if there are no pending items. That is not the number that the method intends to return. Feel free to reopen the conversation if this requires more discussion.

Reply:
@JatinSanghvi Makes sense if that is how the value is used. If the value was used as "This is the max number of instances we can scale up to" that's when I saw an issue with the filtering, because really the max number of instances is the number of states/partitions, regardless of pending changes.